src/MarshalMiddlewareTrait.php

```diff
@@ -35,52 +38,55 @@
      * - If no callable is created, an exception is thrown.
      *
      * @param mixed $middleware
+     * @param Router\RouterInterface $router
+     * @param ResponseInterface $responsePrototype
      * @param null|ContainerInterface $container
-     * @param bool $forError Whether or not generated middleware is intended to
-     *     represent error middleware; defaults to false.
-     * @return callable
+     * @return ServerMiddlewareInterface
      * @throws Exception\InvalidMiddlewareException
      */
-    private function prepareMiddleware($middleware, ContainerInterface $container = null, $forError = false)
-    {
-        if (is_callable($middleware)) {
+    private function prepareMiddleware(
+        $middleware,
+        Router\RouterInterface $router,
+        ResponseInterface $responsePrototype,
+        ContainerInterface $container = null
+    ) {
```

Comment:
I modified the signature to require passing the router and response prototype; this prevents the trait from relying on instance state.